pkg/registration/hub/gc/controller.go

```diff
@@ -0,0 +1,155 @@
+package gc
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorhelpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes"
+	rbacv1listers "k8s.io/client-go/listers/rbac/v1"
+	"k8s.io/client-go/metadata"
+	"k8s.io/klog/v2"
+
+	clientset "open-cluster-management.io/api/client/cluster/clientset/versioned"
+	informerv1 "open-cluster-management.io/api/client/cluster/informers/externalversions/cluster/v1"
+	clusterv1listers "open-cluster-management.io/api/client/cluster/listers/cluster/v1"
+	worklister "open-cluster-management.io/api/client/work/listers/work/v1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+
+	"open-cluster-management.io/ocm/pkg/common/patcher"
+	"open-cluster-management.io/ocm/pkg/common/queue"
+)
+
+type gcReconcileOp int
+
+const (
+	gcReconcileRequeue gcReconcileOp = iota
+	gcReconcileStop
+	gcReconcileContinue
+)
+
+var (
+	addonGvr = schema.GroupVersionResource{Group: "addon.open-cluster-management.io",
+		Version: "v1alpha1", Resource: "managedclusteraddons"}
+	workGvr = schema.GroupVersionResource{Group: "work.open-cluster-management.io",
+		Version: "v1", Resource: "manifestworks"}
+)
+
+// gcReconciler is an interface for reconcile cleanup logic after cluster is deleted.
+// clusterName is from the queueKey, cluster may be nil.
+type gcReconciler interface {
+	reconcile(ctx context.Context, cluster *clusterv1.ManagedCluster) (gcReconcileOp, error)
+}
+
+type GCController struct {
+	clusterLister  clusterv1listers.ManagedClusterLister
+	clusterPatcher patcher.Patcher[*clusterv1.ManagedCluster, clusterv1.ManagedClusterSpec, clusterv1.ManagedClusterStatus]
+	gcReconcilers  []gcReconciler
+}
+
+// NewGCController ensures the related resources are cleaned up after cluster is deleted
+func NewGCController(
+	clusterRoleLister rbacv1listers.ClusterRoleLister,
+	clusterRoleBindingLister rbacv1listers.ClusterRoleBindingLister,
+	roleBindingLister rbacv1listers.RoleBindingLister,
+	clusterInformer informerv1.ManagedClusterInformer,
+	manifestWorkLister worklister.ManifestWorkLister,
+	clusterClient clientset.Interface,
+	kubeClient kubernetes.Interface,
+	metadataClient metadata.Interface,
+	eventRecorder events.Recorder,
+	gcResourceList []string,
+) factory.Controller {
+	clusterPatcher := patcher.NewPatcher[
+		*clusterv1.ManagedCluster, clusterv1.ManagedClusterSpec, clusterv1.ManagedClusterStatus](
+		clusterClient.ClusterV1().ManagedClusters())
+
+	gcResources := []schema.GroupVersionResource{}
+	if len(gcResourceList) == 0 {
```

Comment:
nit: can we give a default value to `GCResourceList` instead of this?

Reply:
the gcResourceList is from the flag `gc-resource-list` with the default values addon and manifetworks.